### PR TITLE
Normalize private symbols in persistent cache keys

### LINKED ISSUE #38828: Persistent compilation cache key changes when private symbols are renumbered by a DCE'd helper (cosmetic @clip_0→@clip_1 rename → spurious recompile)
### Summary
Two jitted programs that produce **byte-identical pretty-printed jaxprs** and semantically-identical StableHLO — differing only in the auto-generated private-symbol *numbers* (`@clip_0` vs `@clip_1`) — get **different persistent compilation cache keys**, so the second run misses the cache and recompiles. The trigger is that one extra helper is emitted during lowering and then DCE'd away; even though it leaves no trace in the final module, it has already bumped MLIR's per-module symbol-uniquing counter, shifting every later private symbol by one.

### Mechanism
- A primitive lowered via the "emit as function" path inserts a `func.func private @name` into the module's `SymbolTable`; the uniquing counter is a single monotonic counter per module, shared across base names.
- If some equation causes **one extra** such helper to be emitted (which then gets inlined / DCE'd before the final module), the counter has still advanced by one, so every subsequently-emitted private symbol is named `_N+1` instead of `_N`. The final modules are structurally identical except for this uniform `+1` on the private-symbol suffixes.
- The persistent cache key hashes the **serialized module**, which carries those symbol names. The only normalization applied before hashing is the MLIR pass `builtin.module(strip-debuginfo)` (`cache_key._hash_computation` → `_canonicalize_ir`), which strips *locations* but **not** symbol *names*. So the `+1` rename changes the hash → different key → cache miss.

### Minimal repro (CPU, jax/jaxlib 0.9.0.1) — verified
`threefry_partitionable(False)` here is just a deterministic way to inject one extra helper that is then DCE'd; it is not the point — any source of an extra-then-removed helper produces the same renumbering.
```python
import numpy as np
import jax, jax.numpy as jnp
from jax._src import compilation_cache, compiler
from jax._src import xla_bridge as xb

def make(split):
    def train_step(x, y):
        a = x + 1
        if split:
            with jax.threefry_partitionable(False):   # injects one extra helper that is then DCE'd
                b = x + 2
        else:
            b = x + 2
        return a + b + jnp.clip(x, 0, 1), jnp.clip(y, 0, 1)
    train_step.__name__ = train_step.__qualname__ = "train_step"
    return train_step

def lowered(split):
    return jax.jit(make(split)).lower(jnp.ones((2,), jnp.float32), jnp.ones((3,), jnp.float32))

def cache_key(lo):
    backend = xb.get_backend("cpu")
    co = compiler.get_compile_options(1, 1, backend=backend)
    return compilation_cache.get_cache_key(lo.compiler_ir("stablehlo"),
                                           np.array([backend.local_devices()[0]]), co, backend)

a, b = lowered(False), lowered(True)
# jaxprs pretty-print identical:
assert jax.make_jaxpr(make(False))(jnp.ones((2,)), jnp.ones((3,))).pretty_print() == \
       jax.make_jaxpr(make(True))(jnp.ones((2,)), jnp.ones((3,))).pretty_print()
# ... but cache keys differ (StableHLO has @clip_0 vs @clip_1):
print("cache_keys_equal:", cache_key(a) == cache_key(b))   # -> False
```
Output: `cache_keys_equal: False`. `b`'s StableHLO is `a`'s with `@clip_0`→`@clip_1`; the extra helper bumped the SymbolTable counter and was then DCE'd (it does not appear in either final module).

We also hit this in practice on larger multi-host programs, where the same uniform `+1` private-symbol rename appears across otherwise-identical StableHLO and causes a full recompile on cache lookup.

### Ask
Could the persistent cache key be made **invariant to cosmetic private-symbol names** — e.g. renumber private symbols by definition order (or strip the numeric suffixes) in `_canonicalize_ir`, alongside the existing `strip-debuginfo` pass, before hashing? That makes semantically-identical programs key identically regardless of incidental symbol renumbering.

### Config
jax/jaxlib 0.9.0.1; repro reproduces on CPU.

### PR DESCRIPTION
**Pull request by vishalvinjamuri27**
Fixes #38828.

Persistent compilation cache keys hash serialized canonicalized StableHLO. The existing canonicalization strips debug info but preserves private helper symbol names, so semantically identical modules can get different keys when private helpers are cosmetically renumbered.

This change canonicalizes top-level private symbol names before hashing by renaming them deterministically in definition order and updating symbol uses through MLIR's SymbolTable. Public symbols are left unchanged. The rename goes through temporary names to avoid collisions when an existing private symbol already has a generated canonical name.

Test plan:
- python3 -m pytest tests/cache_key_test.py::CacheKeyTest::test_identical_computations_different_private_symbol_names -q
- python3 -m pytest tests/cache_key_test.py -q
- python3 -m pytest tests/cache_key_test.py tests/compilation_cache_test.py -q

### CODE REVIEW

**Comment by hawkinsp** [comment]
I'm not convinced this solves a problem that actually exists. We'd need a test case that fails without this change to demonstrate that it's fixing something.
